src/main/io/gps.c

```diff
@@ -573,110 +755,393 @@ static void ubloxSetMessageRate(uint8_t messageClass, uint8_t messageID, uint8_t
+#if !defined(GPS_NMEA_TX_ONLY)
+    uint32_t now;
+#endif
+    DEBUG_SET(DEBUG_GPS_CONNECTION, 4, gpsData.state);
+    DEBUG_SET(DEBUG_GPS_CONNECTION, 5, gpsData.state_position);
+    switch (gpsData.state) {
+        case GPS_STATE_INITIALIZING:
+#if !defined(GPS_NMEA_TX_ONLY)
+           now = millis();
+           if (cmp32(now, gpsData.state_ts) < 1000) {
+               return;
+           }
+           gpsData.state_ts = now;
+            switch (gpsData.state_position) {
+                case 0: // first run after bootup
+                    serialSetBaudRate(gpsPort, 4800);
+                    gpsData.state_position++;
+                    break;
+                case 1: // second run
+                    // print the init string for the baudrate we want to be at
+                    serialPrint(gpsPort, "$PSRF100,1,115200,8,1,0*05\r\n");
+                    gpsData.state_position++;
+                    break;
+                default: 
+                    // we're now (hopefully) at the correct rate, next state should switch to it
+                    // TO DO: Check that this actually works, it seems broken at present
+                    gpsData.state_position = 0;
+                    gpsSetState(GPS_STATE_CHANGE_BAUD);
+                    break;
+            }
             break;
+#endif
+        case GPS_STATE_CHANGE_BAUD:
+#if !defined(GPS_NMEA_TX_ONLY)
+            // wait a short time between sending commands
+            // note that no commands are sent to request the packets we need
+            now = millis();
+            if (cmp32(now, gpsData.state_ts) < 500) {
+                return;
+            }
+            gpsData.state_ts = now;
+
+            if (gpsData.state_position < 1) { // first run after boot up 
+                serialSetBaudRate(gpsPort, baudRates[gpsInitData[gpsData.baudrateIndex].baudrateIndex]);
+                gpsData.state_position++;
+            } else if (gpsData.state_position < 2) {
+                // *** this message also appears to fail ***//
+                // NMEA reports back at whatever speed the module is configured to send at, not 5Hz
+                serialPrint(gpsPort, "$PSRF103,00,6,00,0*23\r\n"); // set GGA rate to 5Hz
+                gpsData.state_position++;
+            } else if (gpsData.state_position < 3) {
+                // special initialization for NMEA ATGM336 and similar GPS recivers - should be done only once
+                if (!atgmRestartDone) {
+                    atgmRestartDone = true;
+                    serialPrint(gpsPort, "$PCAS02,100*1E\r\n");  // 10Hz refresh rate
+                    serialPrint(gpsPort, "$PCAS10,0*1C\r\n");    // hot restart
+                }
+                gpsData.state_position++;
+            } else if (gpsData.state_position < 4) {
+                // try to disable UBX NAV-PVT and NAV-SAT messages via ValSet
+                ubloxSetMessageRateValSet(CFG_MSGOUT_UBX_NAV_PVT_UART1, 0);
```

Comment:
Ah... yes, and no.
NMEA is a total mess; obviously it needs a ton of work.  I should just leave it alone as being outside the scope of this PR.
It seems that gps.c message parser will attempt to parse both NMEA and UBlox incoming data strings, all the time, regardless of the user's selection for one or the other.   
I only know this because putting the UBlox message in this location definitely stops UBLox messages, I can see that while monitoring the traffic from the module, after making the user selection to NMEA.
I will put these UBlox specific commands into `#ifdef Ublox` wrappers.
A later re-factor should make it so that when the user selects UBlox we do not attempt to interrogate and parse NMEA packets.

Reply:
Problem is that you may send binary messages to non-ublox GPS. 
There should be text/nmea equivalent.
But it is better to handle it in another PR